### ISSUE TITLE
Questions regarding reconstruction probability / loss functions with VariationalAutoencoder layer

### ISSUE THREAD

**Issue by skeydan** [deeplearning4j/deeplearning4j#3924]
Hi,

I have problems getting VAE to work with a dataset, other than MNIST, where most of the variables are 1 hot encoded and a few are continuous, scaled to be between between 0 and 1 .

In Keras, where there is no variational layer, I am using the simple architecture from the example code (see ) and have found that the best result is obtained with MSE as the reconstruction error (even though all values lie between 0 and 1). 
So my loss function in Keras is MSE plus KL divergence.

In DL4J, I guess I should then use a Gaussian reconstruction distribution, but by trial-and-error I haven't gotten any satisfying results yet (it seemed that tanh was best).
Do you perhaps have a recommendation which activation to choose?

Also, if I understand correctly I can either use a normal loss function (which is then used for training and scoring the test set) or a reconstruction distribution. In that case, what is used for calculating the loss, i.e., what is the score that is reported?
(Originally I thought that maybe one could use a normal loss function for training, and then use reconstruction probability for testing? But that doesn't seem to be possible then?)

Finally, my reconstruction log probabilities are around -50 to -60 on average, per class. From your experience, does this look plausible?

Sorry for these questions... Would be great to get any feedback on this!
Thank you very much in advance
Sigrid

**Comment by AlexDBlack** [comment]
For data in the range 0 to 1 (including 0/1 only): I'd go with sigmoid activation and bernoulli reconstruction distribution.
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/variational/BernoulliReconstructionDistribution.java

MSE + gaussian might work too, but it's not really an appropriate distribution.

You are correct that you can alternatively use a normal loss function. The original Kingma & Welling paper used reconstruction distributions (i.e., reconstructions are probabilistic), but I've seen standard (deterministic) loss functions used also in other implementations (but not in the literature). Standard loss functions should work fine in theory though.

Hard to say as to the exact log probabilities - really depends on the data and choice of other parameters.

**Comment by skeydan** [comment]
Thank you!
